Cookbook/Proposals/ComposableReducers.md

```diff
@@ -0,0 +1,250 @@
+
+The basic idea can be found in @mbrandonw ’s talk:
+
+[Brandon Williams \- Composable Reducers & Effects Systems \- YouTube](https://www.youtube.com/watch?v=QOIigosUNGU)
+
+## Proposed solution
+
+### Functional `Lens` & `Prism`
+
+```swift
+/// For accessing struct members.
+/// e.g. Whole = whole struct (members), Part = partial member
+struct Lens<Whole, Part> {
+    let get: (Whole) -> Part
+    let set: (Whole, Part) -> Whole
+}
+
+/// For accessing enum cases.
+/// e.g. Whole = all possible enum cases, B = partial case
+struct Prism<Whole, Part> {
+    let preview: (Whole) -> Part?
+    let review: (Part) -> Whole
+}
+```
+
+- `Lens` is a pair of "getter" and "setter" (similar to `WritableKeyPath<A, B>`, but more composable)
+- `Prism` is a pair of:
+    - `preview` (tryGet): Tries to get an associated value of particular enum case from whole enum cases, which is failurable
+    - `review` (inject): Creates whole enum from particular case (i.e. `case` as enum constructor)
+
+While `Lens` is useful for traversing `struct` members, `Prism` is useful for traversing `enum` cases.
+Because in ReactiveFeedback, `State` is normally defined as `struct` and `Event` is `enum`,
+we need both features to be able to transform `reducer` and `feedback` into arbitrary structure.
+
+### `Reducer`
+
+```swift
+struct Reducer<Action, State> {
+    let reduce: (Action, State) -> State
+
+    init(_ reduce: @escaping (Action, State) -> State) {
+        self.reduce = reduce
+    }
+
+    /// Zero value for `+`.
+    static var empty: Reducer {
+        return Reducer { _, s in s }
+    }
+
+    // Append operator, just like `+`.
+    static func <> (lhs: Reducer, rhs: Reducer) -> Reducer {
+        return Reducer { action, state in
+            rhs.reduce(action, lhs.reduce(action, state))
+        }
+    }
+}
+```
+
+`Reducer` is a wrapper type around `reduce: (Action, State) -> State` function that conforms to `Monoid` (has "zero" and "+") to combine 2 reducers into one.
+
+By using this `append`ing capability, we can create more complex `Reducer` from splitted `SubReducer`s.
+
+### `Reducer` lifting from `SubState` / `SubAction`
+
+However, `SubReducer`s don't normally have the same type with the others, even with the `(Main)Reducer` type.
+
+For example,
+
+- Main screen: `MainReducer = Reducer<MainAction, MainState>`
+    - Component 1: `SubReducer1 = Reducer<Sub1Action, Sub1State>`
+    - Component 2: `SubReducer2 = Reducer<Sub2Action, Sub2State>`
+    - ...
+
+To convert `SubReducer1` and `SubReducer2` types into `MainReducer` (so that they can be combined using `<>`), we can use the following `lift` functions:
+
+```swift
+extension Reducer {
+    /// `Reducer<Action, SubState> -> `Reducer<Action, State>`
+    func lift<SuperState>(state lens: Lens<SuperState, State>) -> Reducer<Action, SuperState> {
+        return Reducer<Action, SuperState> { action, superState in
+            lens.set(superState, self.reduce(action, lens.get(superState)))
```

Comment:
the lens looks like WritableKeyPath, maybe we should use just it or Lens allows more things?

```swift
func lift<SuperState>(state keyPath: WritableKeyPath<SuperState, State>) -> Reducer<Action, SuperState> {
  return Reducer<Action, SuperState> { action, superState in
    superState[keyPath] = self.reduce(action, superState[keyPath])
  }
}
```

Reply:
We can use `WritableKeyPath` to always make `Lens` :

```swift
func lens<Whole, Part>(_ keyPath: WritableKeyPath<Whole, Part>) -> Lens<Whole, Part> {
    return Lens<Whole, Part>(
        get: { $0[keyPath: keyPath] },
        set: { whole, part in
            var whole = whole
            whole[keyPath: keyPath] = part
            return whole
        }
    )
}

let reducer = subReducer.lift(state: lens(\.state))
```

My preference is always convert to `Lens` whenever possible.
But I don't mind always having `WritableKeyPath` overloads.